lib/translator/inquiry.cc

```diff
@@ -13,10 +13,90 @@
 // limitations under the License.
 
 #include "inquiry.h"
+
 #include "string.h"
 
 namespace translator {
 
+// anonymous namespace for helper functions and constant expressions
+namespace {
+constexpr uint8_t identifier_length_eui64 = 0x8;
+constexpr uint8_t identifier_length_nguid = 0x10;
+
+// Builds the IdentificationDescriptor according to 6.1.4.5 of the translation
+// spec:
+// https://nvmexpress.org/wp-content/uploads/NVM_Express_-_SCSI_Translation_Reference-1_5_20150624_Gold.pdf
+// This only builds the constant part of the Identification Descriptor. The
+// caller is responsible for building the identifier.
+scsi_defs::IdentificationDescriptor BuildIdentificationDescriptor(
+    const nvme_defs::IdentifyNamespace &identify_namespace_data) {
+  // check if nonzero
+  bool nguid_nz =
+      identify_namespace_data.nguid[0] || identify_namespace_data.nguid[1];
+  bool eui64_nz = identify_namespace_data.eui64;
+  uint8_t identifier_length;
+
+  if (nguid_nz) {
+    /*
+    Shall be set to 08h if the EUI-64 field is populated from the NVMe EUI64
+    field.
+    */
+    identifier_length = identifier_length_nguid;
+
+    uint8_t identification_descriptor_identifier[16];
+
+    memcpy(&identification_descriptor_identifier,
+           &identify_namespace_data.nguid,
+           sizeof(identification_descriptor_identifier));
+
+  } else if (eui64_nz) {
+    /*
+    Shall be set to 10h if the EUI-64 field is populated from the NVMe NGUID
+    field.
+    */
+    identifier_length = identifier_length_eui64;
+
+    uint64_t identification_descriptor_identifier =
+        identify_namespace_data.eui64;
+  } else {
+    // function not supported as both NGUID and EUI64 fields are zero
+    DebugLog("Both NGUID and EUI-64 fields are zero in IdentityNamespaceData");
+  }
+
+  scsi_defs::IdentificationDescriptor identification_descriptor = {
+
+      /*
+      Shall be set to 0h. PIV field shall indicate this field is reserved as no
+      specific protocol to be identified.
+      */
+      .protocol_identifier = 0x0,
+
+      /*
+      Shall be set to 1h indicating associated fields are in binary format.
+      */
+      .code_set = 0x1,
+
+      /*
+      Shall be set to 0b indicating PROTOCOL IDENTIFIER field is reserved.
+      */
+      .piv = 0,
+
+      /*
+      Shall be set to 00b indicating DESIGNATOR field is associated with logical
+      unit.
+      */
+      .association = 0b00,
+
+      /*
+      Shall be set to 2h indicating EUI-64 based identifier.
+      */
+      .identifier_type = 0x2,
```

Comment:
It is good practice to have named variables for constants in your code (even if they are used only one or a few times). This makes code easier to read, and ensures that multiple uses of the same constant are guaranteed to be consistent (checking for these values in your tests for example).

Constants like `0b00 `, or `0x2 ` don't mean anything to the average reader, so having descriptive names for them make the piece of code much easier to understand.

Reply:
That's a great suggestion. I realized that we can have enums for all these fields and therefore I added them. Let me know what you think about that